File: app/src/test/java/com/slapshotapps/dragonshockey/activities/admin/viewmodels/AdminStatsViewModelTest.java
package com.slapshotapps.dragonshockey.activities.admin.viewmodels;

import com.slapshotapps.dragonshockey.models.PlayerPosition;
import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class AdminStatsViewModelTest {

    @Test
    public void testPlayerNameWhenNull(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator(null, PlayerPosition.FORWARD);

        assertThat(viewModel.getPlayerName(), is(""));
    }

    @Test
    public void testPlayerNameNonNull(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD);

        assertThat(viewModel.getPlayerName(), is("Test"));
    }

    @Test
    public void testDataIsNotDirtyWithNoChange(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD);

        assertFalse(viewModel.isDirty());
    }

    @Test
    public void testDataIsDirtyWhenGoalsUpdated(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD);

        assertFalse(viewModel.isDirty());

        viewModel.setGoals("");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getGoals(), is("0"));

        viewModel.setGoals("5");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getGoals(), is("5"));
    }

    @Test
    public void testDataIsDirtyWhenAssistsUpdated(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD);

        assertFalse(viewModel.isDirty());

        viewModel.setAssists("");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getGoals(), is("0"));

        viewModel.setAssists("3");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getAssists(), is("3"));
    }

    @Test
    public void testDataIsDirtyWhenPIMUpdated(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD);

        assertFalse(viewModel.isDirty());

        viewModel.setPenaltyMinutes("");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getPenaltyMinutes(), is("0"));

        viewModel.setPenaltyMinutes("15");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getPenaltyMinutes(), is("15"));
    }

    @Test
    public void testDataIsDirtyWhenGoalsAgainstUpdated(){
        AdminStatsViewModel viewModel = adminStatsViewModelCreator("Test", PlayerPosition.GOALIE);

        assertFalse(viewModel.isDirty());

        viewModel.setGoalsAgainst("");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getGoalsAgainst(), is("0"));

        viewModel.setGoalsAgainst("5");
        assertTrue(viewModel.isDirty());
        assertThat(viewModel.getGoalsAgainst(), is("5"));
    }

    private AdminStatsViewModel adminStatsViewModelCreator(String name, PlayerPosition position){
        return new AdminStatsViewModel(name, 0,0,1,56,
            true, 0, position, 1);
    }
}

File: app/src/test/java/com/slapshotapps/dragonshockey/activities/admin/viewmodels/AdminStatsViewModelTest.kt
package com.slapshotapps.dragonshockey.activities.admin.viewmodels

import com.slapshotapps.dragonshockey.models.PlayerPosition
import org.junit.Before
import org.junit.Test

import org.hamcrest.CoreMatchers.`is`
import org.junit.Assert.*

class AdminStatsViewModelTest {

    @Test
    fun testPlayerNameWhenNull() {
        val viewModel = adminStatsViewModelCreator(null, PlayerPosition.FORWARD)

        assertThat(viewModel.playerName, `is`(""))
    }

    @Test
    fun testPlayerNameNonNull() {
        val viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD)

        assertThat(viewModel.playerName, `is`("Test"))
    }

    @Test
    fun testDataIsNotDirtyWithNoChange() {
        val viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD)

        assertFalse(viewModel.isDirty)
    }

    @Test
    fun testDataIsDirtyWhenGoalsUpdated() {
        val viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD)

        assertFalse(viewModel.isDirty)

        viewModel.goals = ""
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.goals, `is`("0"))

        viewModel.goals = "5"
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.goals, `is`("5"))
    }

    @Test
    fun testDataIsDirtyWhenAssistsUpdated() {
        val viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD)

        assertFalse(viewModel.isDirty)

        viewModel.assists = ""
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.goals, `is`("0"))

        viewModel.assists = "3"
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.assists, `is`("3"))
    }

    @Test
    fun testDataIsDirtyWhenPIMUpdated() {
        val viewModel = adminStatsViewModelCreator("Test", PlayerPosition.FORWARD)

        assertFalse(viewModel.isDirty)

        viewModel.penaltyMinutes = ""
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.penaltyMinutes, `is`("0"))

        viewModel.penaltyMinutes = "15"
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.penaltyMinutes, `is`("15"))
    }

    @Test
    fun testDataIsDirtyWhenGoalsAgainstUpdated() {
        val viewModel = adminStatsViewModelCreator("Test", PlayerPosition.GOALIE)

        assertFalse(viewModel.isDirty)

        viewModel.goalsAgainst = ""
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.goalsAgainst, `is`("0"))

        viewModel.goalsAgainst = "5"
        assertTrue(viewModel.isDirty)
        assertThat(viewModel.goalsAgainst, `is`("5"))
    }

    private fun adminStatsViewModelCreator(name: String?, position: PlayerPosition): AdminStatsViewModel {
        return AdminStatsViewModel(name, 0, 0, 1, 56,
                true, 0, position, 1)
    }
}